File: civsim/src/CivSim.Core/Geo/LocalFrame.cs
using System;

namespace CivSim.Core.Geo
{
    /// <summary>
    /// Flat metric frame centred on Seoul City Hall: X east, Z north, metres.
    /// Equirectangular approximation; within the 38 × 32 km play area the error is a few metres, far below cell size.
    /// </summary>
    public sealed class LocalFrame
    {
        public const double EarthRadiusM = 6371008.8;

        public static readonly LocalFrame Seoul = new LocalFrame(37.5665, 126.9780);

        /// <summary>Play-area bounding box from the plan (§1.3).</summary>
        public const double SeoulMinLon = 126.76, SeoulMaxLon = 127.19, SeoulMinLat = 37.42, SeoulMaxLat = 37.71;

        public double OriginLatDeg { get; }
        public double OriginLonDeg { get; }
        private readonly double _cosLat;

        public LocalFrame(double originLatDeg, double originLonDeg)
        {
            OriginLatDeg = originLatDeg;
            OriginLonDeg = originLonDeg;
            _cosLat = Math.Cos(originLatDeg * Math.PI / 180.0);
        }

        public void ToLocal(double latDeg, double lonDeg, out double xEast, out double zNorth)
        {
            double k = EarthRadiusM * Math.PI / 180.0;
            xEast = (lonDeg - OriginLonDeg) * k * _cosLat;
            zNorth = (latDeg - OriginLatDeg) * k;
        }

        public void ToLatLon(double xEast, double zNorth, out double latDeg, out double lonDeg)
        {
            double k = EarthRadiusM * Math.PI / 180.0;
            lonDeg = OriginLonDeg + xEast / (k * _cosLat);
            latDeg = OriginLatDeg + zNorth / k;
        }
    }
}

File: civsim/web/src/core/geo/localFrame.ts
/**
 * Flat metric frame centred on Seoul City Hall: +X east, +Z north, metres.
 * Equirectangular approximation; across the 38 x 32 km play area the error stays under a few metres,
 * far below the 100 m simulation cell.
 */
export const ORIGIN_LAT = 37.5665;
export const ORIGIN_LON = 126.978;
export const M_PER_DEG = 111194.92664455873; // 2*pi*R/360, R = 6371008.8 m

const COS_LAT = Math.cos((ORIGIN_LAT * Math.PI) / 180);

export function lonLatToLocal(lon: number, lat: number): { x: number; z: number } {
  return { x: (lon - ORIGIN_LON) * M_PER_DEG * COS_LAT, z: (lat - ORIGIN_LAT) * M_PER_DEG };
}

export function localToLonLat(x: number, z: number): { lon: number; lat: number } {
  return { lon: ORIGIN_LON + x / (M_PER_DEG * COS_LAT), lat: ORIGIN_LAT + z / M_PER_DEG };
}

export function formatLonLat(lon: number, lat: number): string {
  const ns = lat >= 0 ? 'N' : 'S';
  const ew = lon >= 0 ? 'E' : 'W';
  return `${Math.abs(lat).toFixed(4)}°${ns} ${Math.abs(lon).toFixed(4)}°${ew}`;
}
